fish-rust/src/abbrs.rs

```diff
@@ -0,0 +1,479 @@
+    }
+
+    /// Rename an abbreviation. This asserts that the old name is used, and the new name is not; the
+    /// caller should check these beforehand with has_name().
+    pub fn rename(&mut self, old_name: &wstr, new_name: &wstr) {
+        let erased = self.used_names.remove(old_name);
+        let inserted = self.used_names.insert(new_name.to_owned());
+        assert!(
+            erased && inserted,
+            "Old name not found or new name already present"
+        );
+        for abbr in self.abbrs.iter_mut() {
+            if abbr.name == old_name {
+                abbr.name = new_name.to_owned();
+                break;
+            }
+        }
+    }
+
+    /// Erase an abbreviation by name.
+    /// \return true if erased, false if not found.
+    pub fn erase(&mut self, name: &wstr) -> bool {
+        let erased = self.used_names.remove(name);
+        if !erased {
+            return false;
+        }
+        for (index, abbr) in self.abbrs.iter().enumerate().rev() {
+            if abbr.name == name {
+                self.abbrs.remove(index);
+                return true;
+            }
+        }
+        panic!("Unable to find named abbreviation");
+    }
+
+    /// \return true if we have an abbreviation with the given name.
+    pub fn has_name(&self, name: &wstr) -> bool {
+        self.used_names.contains(name)
+    }
+
+    /// \return a reference to the abbreviation list.
+    pub fn list(&self) -> &[Abbreviation] {
+        &self.abbrs
+    }
+}
+
+/// \return the list of replacers for an input token, in priority order, using the global set.
+/// The \p position is given to describe where the token was found.
+fn abbrs_match_ffi(token: &CxxWString, position: abbrs_position_t) -> Vec<abbrs_replacer_t> {
+    with_abbrs(|set| set.r#match(&token.from_ffi(), position.into()))
+        .into_iter()
+        .map(|r| r.into())
+        .collect()
+}
+
+fn abbrs_has_match_ffi(token: &CxxWString, position: abbrs_position_t) -> bool {
+    with_abbrs(|set| set.has_match(&token.from_ffi(), position.into()))
+}
+
+fn abbrs_list_ffi() -> Vec<abbreviation_t> {
+    with_abbrs(|set| -> Vec<abbreviation_t> {
+        let list = set.list();
+        let mut result = Vec::with_capacity(list.len());
+        for abbr in list {
+            result.push(abbreviation_t {
+                key: abbr.key.to_ffi(),
+                replacement: abbr.replacement.to_ffi(),
+                is_regex: abbr.is_regex(),
+            })
+        }
+
+        result
+    })
+}
+
+fn abbrs_get_set_ffi<'a>() -> Box<GlobalAbbrs<'a>> {
+    let abbrs_g = abbrs.lock().unwrap();
+    Box::new(GlobalAbbrs { g: abbrs_g })
+}
+
+fn abbrs_replacement_from_ffi(
+    range: SourceRange,
+    text: &CxxWString,
+    set_cursor_marker: &CxxWString,
+    has_cursor_marker: bool,
+) -> abbrs_replacement_t {
+    let mut cursor_marker = None;
+    if has_cursor_marker {
+        let _ = cursor_marker.insert(set_cursor_marker.from_ffi());
```

Comment:
nit: Id write this as `let cursor_marker = if has_cursor_marker { Some(...) } else { None }`